special-pages/pages/duckplayer/app/features/error-detection.js

```diff
@@ -0,0 +1,139 @@
+import { YOUTUBE_ERROR_EVENT, YOUTUBE_ERRORS } from '../providers/YouTubeErrorProvider';
+
+/**
+ * @typedef {import("./iframe").IframeFeature} IframeFeature
+ * @typedef {import('../../types/duckplayer').YouTubeError} YouTubeError
+ * @typedef {import('../../types/duckplayer').DuckPlayerPageSettings['customError']} CustomErrorOptions
+ */
+
+/**
+ * Detects YouTube errors based on DOM queries
+ *
+ * @implements IframeFeature
+ */
+export class ErrorDetection {
+    /** @type {HTMLIFrameElement} */
+    iframe;
+
+    /** @type {CustomErrorOptions} */
+    options;
+
+    /**
+     * @param {CustomErrorOptions} options
+     */
+    constructor(options) {
+        this.options = options;
+    }
+
+    /**
+     * @param {HTMLIFrameElement} iframe
+     */
+    iframeDidLoad(iframe) {
+        this.iframe = iframe;
+
+        if (!this.options || !this.options.signInRequiredSelector) {
+            console.log('Missing Custom Error options');
+            return null;
+        }
+
+        const documentBody = iframe.contentWindow?.document?.body;
+        if (documentBody) {
+            // Check if iframe already contains error
+            if (this.checkForError(documentBody)) {
+                const error = this.getErrorType();
+                window.dispatchEvent(new CustomEvent(YOUTUBE_ERROR_EVENT, { detail: { error } }));
+
+                return null;
+            }
+
+            // Create a MutationObserver instance
+            const observer = new MutationObserver(this.handleMutation.bind(this));
+
+            // Start observing the iframe's document for changes
+            observer.observe(documentBody, {
```

Comment:
In the case where you setup this observer, you should return a cleanup funtion to disconnect it. `iframeDidLoad` accepts a callback that it will call